### PR TITLE
Adding support to inject headers and intercept Http outgoing requests

### PR DESCRIPTION
**Pull request by brahmnes**
using DiagnosticSource. This frees the user from having to use the
profiler.  Also introduced HttpProcessing as the base class for
FrameworkHttpProcessing and ProfilerHttpProcessing to unify the logic
from both paths in handling correlation headers and dependency telemetry
handling.

A temporary HttpHandlerDiagnosticListener implementation is added to
make all this work without a new version of DiagnosticSource nuget. This
class is currently under reviewed in the corefx repo and will be part of
DiagnosticSource. Once that happens, we will remove this temporary
implementation of HttpHandlerDiagnosticListener.

### CODE REVIEW

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+
+            servicePointTableField.SetValue(null, newTable);
+        }
+
+        /// <summary>
+        /// Initializes all the reflection objects it will ever need. Reflection is costly, but it's better to take
+        /// this one time performance hit than to get it multiple times later, or do it lazily and have to worry about
+        /// threading issues. If Initialize has been called before, it will not doing anything.
+        /// </summary>
+        private void Initialize()
```
Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: I think you should use a static constructor here and avoid the need for initializationLock.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    /// <summary>
+    /// A HttpHandlerDiagnosticListener is a DiagnosticListener for .NET 4.5/4.6 where HttpClient
+    /// doesn't have a DiagnosticListener built-in. This class is not used for .NET Core because
+    /// HttpClient in .NET Core already emits DiagnosticSource events. This class compensates for
+    /// that in .NET 4.5/4.6. HttpHandlerDiagnosticListener has no public constructor, but it has a
+    /// singleton style Initialize method to control its creation.
+    /// To use this, the application just needs to call HttpHandlerDiagnosticListener.Initialize(), and this
+    /// will register itself with the DiagnosticListener's all listeners collection.
+    /// </summary>
+    internal class HttpHandlerDiagnosticListener : DiagnosticListener
```
Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: sealed

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    {
+        internal static HttpHandlerDiagnosticListener SingletonInstance = new HttpHandlerDiagnosticListener();
```
Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: readonly

Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: You could call this just ```Instance```. The "Singleton" prefix doesn't really add anything that ```static``` doesn't already tell you.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    using System.Reflection;
+    using System.Reflection.Emit;
+
+    /// <summary>
+    /// A HttpHandlerDiagnosticListener is a DiagnosticListener for .NET 4.5/4.6 where HttpClient
+    /// doesn't have a DiagnosticListener built-in. This class is not used for .NET Core because
+    /// HttpClient in .NET Core already emits DiagnosticSource events. This class compensates for
+    /// that in .NET 4.5/4.6. HttpHandlerDiagnosticListener has no public constructor, but it has a
+    /// singleton style Initialize method to control its creation.
+    /// To use this, the application just needs to call HttpHandlerDiagnosticListener.Initialize(), and this
```
Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Is this right? Initialize is private. How does the application call it?

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        internal static HttpHandlerDiagnosticListener SingletonInstance = new HttpHandlerDiagnosticListener();
+
+#region private fields
+        private const string DiagnosticListenerName = "System.Net.Http.Desktop.V4";
+        private const string RequestWriteName = "System.Net.Http.Request";
+        private const string ResponseWriteName = "System.Net.Http.Response";
+        private const string ExceptionEventName = "System.Net.Http.Exception";
+
+        // Fields for reflection
+        private static FieldInfo connectionGroupListField;
```
Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: If you use a static constructor instead of Initialize, then all these static fields should become ```readonly```

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        private static Func<HttpWebRequest, HttpWebResponse> httpResponseAccessor;
+
+        // Fields for controlling initialization of the HttpHandlerDiagnosticListener singleton
+        private object initializationLock = new object();
+        private bool initialized = false;
+#endregion
+
+        /// <summary>
+        /// Prevents a default instance of the <see cref="HttpHandlerDiagnosticListener" /> class from being created.
+        /// Make sure every caller has to create this object via the Initialize method.
```
Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Comment out of date? Seems that (internal) callers just access the singleton, right?

File: Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceSubscriber.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    internal class HttpDiagnosticSourceSubscriber : IObserver<DiagnosticListener>, IDisposable
+    {
+        private HttpDiagnosticSourceListener parent;
+        private IDisposable allListenersSubscription;
+        private IDisposable sourceSubscription;
+        private bool disposed = false;
+
+        internal HttpDiagnosticSourceSubscriber(HttpDiagnosticSourceListener parent)
+        {
+            GC.KeepAlive(HttpHandlerDiagnosticListener.SingletonInstance);
```
Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceSubscriber.cs]: GC,KeepAlive is a bit odd (I don't think the GC will ever collect static fields), but perhaps it captures the intent: This is the place to create the singleton and to hook up its events. Right?

Comment (pharring) [inline Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceSubscriber.cs]: Also, is this bit of code going to migrate into Corefx too? If not, then the SingletonInstance presumably needs to be made public.

**Comment by SergeyKanzhelev** [comment]
>Once that happens, we will remove this temporary
implementation of HttpHandlerDiagnosticListener.

@brahmnes are we talking `2.4.0-beta1` or `2.4..0` timeframe?

**Comment by SergeyKanzhelev** [comment]
Please update `CHANGELOG.md`

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+        /// </summary>
+        public override IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Func<string, object, object, bool> isEnabled)
+        {
+            this.Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+#region private methods
+
+        private static void PrepareReflectionObjects()
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: We need somebody from System.Net to sign off on these changes.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+                    this.initialized = true;
+
+                    try
+                    {
+                        PrepareReflectionObjects();
+                        PerformInjection();
+                    }
+                    catch (Exception)
+                    {
+                        // If anything went wrong, just no-op
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: >// If anything went wrong, just no-o [](start = 24, length = 36)

Please add tracing to troubleshoot why exactly nothing works

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+        {
+            this.Initialize();
+            return base.Subscribe(observer, isEnabled);
+        }
+
+#region private methods
+
+        private static void PrepareReflectionObjects()
+        {
+            // At any point, if the operation failed, it should just throw, including NullReferenceException.
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: >// At any point, if the operation failed, it should just throw, including NullReferenceException. [](start = 12, length = 97)

please minimize the number of thrown exceptions. Ideally make it `TryPrepareReflectionObjects` instead and return the inner class containing all the needed reflection types/delegates

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+
+            // Double checking to make sure we have all the pieces initialized
+            if (connectionGroupListField == null ||
+                connectionGroupType == null ||
+                connectionListField == null ||
+                connectionType == null ||
+                writeListField == null ||
+                httpResponseAccessor == null)
+            {
+                throw new InvalidOperationException("Unable to initialize all required reflection objects");
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: >"Unable to initialize all required reflection objects" [](start = 52, length = 54)

if this class would become the part of Diagnostics Source - I'd suggest to minimize the number of places your use English strings as ultimately all of them will need to be translated. Application Insights at the moment has more relaxed requirements on localization. So maybe just return false.

File: Src/DependencyCollector/Net45/DependencyCollector.Net45.csproj [hunk 12]
```diff
@@ -45,6 +45,7 @@
     <Reference Include="System.Windows.Forms" />
     <Reference Include="System.Xml" />
     <Reference Include="System.Xml.Linq" />
+    <Reference Include="Microsoft.CSharp" />
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Net45/DependencyCollector.Net45.csproj]: >"Microsoft.CSharp" /> [](start = 23, length = 21)

is it needed for `Reflection.Emit`?

File: Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs [hunk 13]
```diff
@@ -1,40 +1,42 @@
...
+    /// received from framework events for HTTP. This class receives and uses two types of events, which
+    /// are EventSource from FrameworkHttpEventListener, and DiagnosticSource from
+    /// HttpDiagnosticSourceListener. The challenge is, the diagnostic source events have the WebRequest
+    /// object which we need for header injection, but the events can be fired multiple times. So you won't
+    /// know if it's the first request, or if it's the last response. The event source events fire at the right
+    /// locations and just once, but they don't have the rich information. This class coordinates both events,
+    /// store information in DependencyTelemetry properly, and fire the telemetry only on the EventSource
+    /// response receive to guarantee it's done just once and at the right time.
+    /// </summary>
+    internal sealed class FrameworkHttpProcessing : HttpProcessing
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs]: >Framework [](start = 26, length = 9)

`Framework` prefix meant "collected via built-in framewrok instrumentation". As opposite to `Profiler` prefix. I'd suggest to differentiate RDD collected via EventSource and Diagnostics Source and has their prefixes separately.

File: Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs [hunk 14]
```diff
@@ -1,40 +1,42 @@
...
     {
         internal CacheBasedOperationHolder TelemetryTable;
         private readonly ApplicationInsightsUrlFilter applicationInsightsUrlFilter;
-        private readonly TelemetryClient telemetryClient;
 
-        internal FrameworkHttpProcessing(TelemetryConfiguration configuration, CacheBasedOperationHolder telemetryTupleHolder)
+        internal FrameworkHttpProcessing(TelemetryConfiguration configuration, CacheBasedOperationHolder telemetryTupleHolder, bool setCorrelationHeaders, ICollection<string> correlationDomainExclusionList, string appIdEndpoint)
+            : base(configuration, SdkVersionUtils.GetSdkVersion("rdd" + RddSource.Framework + ":"), null, setCorrelationHeaders, correlationDomainExclusionList, appIdEndpoint)
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs]: >"rdd" + RddSource.Framework [](start = 64, length = 27)

need new sdk version prefix for diagnostics source

File: Src/DependencyCollector/Shared/Implementation/HttpProcessing.cs [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+#if NET40
+    using Microsoft.ApplicationInsights.Net40;
+#endif
+    using Microsoft.ApplicationInsights.Web.Implementation;
+
+    /// <summary>
+    /// Concrete class with all processing logic to generate RDD data from the callbacks
+    /// received from Profiler instrumentation for HTTP or HTTP EventSource/DiagnosticSource events.   
+    /// </summary>
+    internal abstract class HttpProcessing
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/HttpProcessing.cs]: >abstract [](start = 13, length = 8)

we found this generalization a bad idea. Abstract class starts knowing all sorts of internal details of children classed. In this code review - Profiler and Diagnostics Source may share headers reading and injection logic as well as appending VERB to the name. So perhaps should share the base class as well. Also adding new collectors becoming a very hard task as you may need to refactor how the base class works. Consider adding helper classes instead of the base class.

File: Src/DependencyCollector/Shared/DependencyTrackingTelemetryModule.cs [hunk 16]
```diff
@@ -177,9 +183,12 @@ protected virtual void Dispose(bool disposing)
         private void InitializeForFrameworkEventSource()
```
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/DependencyTrackingTelemetryModule.cs]: >InitializeForFrameworkEventSource [](start = 21, length = 33)

Now you have two implementations  - one that requires reflection and may fail to initialize and another that requires less. I do not see the fall back logic implemented. Am I missing something?

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 1 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: This file is temporary. The real implementation will sit in System.Diagnostics.DiagnosticSource, see this review at:
https://github.com/dotnet/corefx/pull/18101

Once it's in DiagnosticSource dll, this file will be removed. It also makes more sense if you look at how we trigger the initialization from DiagnosticListener, it becomes clear that you don't want to use a static constructor because it should get initialized only if someone is subscribing for listening.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 2 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Sure

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 3 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: I will leave it as is since it passes Stylecop now, and this file will be removed in a week or two (hopefully). In DiagnosticSource, this is named as s_instance and it's consistent with fields in DiagnosticListener

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 4 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: The GC.KeepAlive on the instance field initializes it and calls the constructor, which then inserts itself into DiagnosticListener.AllListeners. Then the caller gets a callback in OnNext for subscribing, which then triggers the initialization. The flow is a little complicating, but that's basically how DiagnosticListener subscription works today.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 5 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Again, it's a initialized only on subscribe, so can't use static constructor here.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 6 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Yes, will update the comment.

File: Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceSubscriber.cs (hunk 7 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceSubscriber.cs]: Yes, this code will go into corefx.
You are correct that GC.KeepAlive is probably not necessary, your assumption is correct.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 8 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: There is a parallel PR in corefx:
https://github.com/dotnet/corefx/pull/18101

Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: To your earlier question, this file will be removed hopefully within in a few weeks. So yes, it's in 2.4.0 timeframe.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 9 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: I probably won't change this here. I will look at corefx to see what tracing they use there

File: Src/DependencyCollector/Shared/DependencyTrackingTelemetryModule.cs (hunk 16 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/DependencyTrackingTelemetryModule.cs]: The FrameworkHttpEventListener path is unchanged. FrameworkHttpProcessing.OnBeginHttpCallback and OnEndHttpCallback are still there. The worst case is, you just get what you get today.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 11 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Will take your suggestion about changing these functions to return bool, so these strings should go away.

File: Src/DependencyCollector/Net45/DependencyCollector.Net45.csproj (hunk 12 above)
Comment (brahmnes) [inline Src/DependencyCollector/Net45/DependencyCollector.Net45.csproj]: It's needed for the "dynamic" keyword

File: Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceListener.cs [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+                        this.httpProcessingFramework.OnRequestSend((HttpWebRequest)propertyBag.Request);
+                        break;
+                    case "System.Net.Http.Response":
+                        this.httpProcessingFramework.OnResponseReceive((HttpWebRequest)propertyBag.Request, (HttpWebResponse)propertyBag.Response);
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                // This shouldn't happen, but just in case it does, we shouldn't break the application.
```
Comment (lmolkova) [inline Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceListener.cs]: as far as i can tell, all exceptions are handled in OnBegin/OnEnd.
However if there is a reason to catch them, consider writing error to the event log

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 9 above)
Comment (SergeyKanzhelev) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: We will need tracing in Application Insights as well. We get support questions on why dependencies weren't collected way too often

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 10 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Addressing it in the corefx repo

File: Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs (hunk 13 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs]: It's the same processing class, you cannot split this between Diagnostic and Framework.

File: Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs (hunk 14 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs]: I don't understand that that means. Can you tell me specifically what I need to change?

File: Src/DependencyCollector/Shared/Implementation/HttpProcessing.cs (hunk 15 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpProcessing.cs]: This is done to minimize changes to the ProfilerHttpProcessing class as much as possible. I believe given the current circumstances, this is probably the best option with lowest chance of regressing tests and functionality and be able to ship on time. If I start refactoring this into a helper class, a whole bunch of references and unit tests would break because the things they reference into ProfilerHttpProcessing have all of a sudden gone missing. To really address it the way you want, much more time is needed with consideration of what our test strategy should be.

File: Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceListener.cs (hunk 17 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceListener.cs]: Adding a call to DependencyCollectorEventSource

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 11 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Actually it looks like this is not an issue in the corefx repo, will keep the exceptions and strings. The exception will be thrown at most once during initialization, so it's not high traffic.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 9 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Will write an event.

File: CHANGELOG.md [hunk 18]
```diff
@@ -4,6 +4,7 @@
 - Report status code for the dependencies failed with non-protocol issue like DNS resolution or SSL shakeup problems.
 - Implemented automatic telemetry correlation: all telemetry reported within the scope of the request is correlated to RequestTelemetry reported for the request.
 - Implemented [Correlation HTTP protocol](https://github.com/lmolkova/correlation/blob/master/http_protocol_proposal_v1.md): default headers to pass Operation Root Id and Parent Id were changed. This version is backward compatible with previously supported headers. 
+- Implemented injection into the HTTP stack to leverage DiagnosticSource to gain access to the WebRequest and WebResponse objects for header injections, without the need of using the profiler.
```
Comment (cijothomas) [inline CHANGELOG.md]: does this work in all .net versions? if not, can you mention it in here?

File: Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs [hunk 19]
```diff
@@ -64,7 +79,20 @@ public void TestHttpPostRequestsAreCollected()
         public void TestHttpRequestsWithQueryStringAreCollected()
         {
             ITelemetry sentTelemetry = null;
-            var channel = new StubTelemetryChannel { OnSend = telemetry => sentTelemetry = telemetry };
+
+            var channel = new StubTelemetryChannel
```
Comment (cijothomas) [inline Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs]: this stubchannel code can be reused without copy pasting for every method.

File: Src/DependencyCollector/Shared.Tests/Implementation/FrameworkHttpProcessingTest.cs [hunk 20]
```diff
@@ -21,17 +25,18 @@ namespace Microsoft.ApplicationInsights.DependencyCollector.Implementation
     [TestClass]
     public sealed class FrameworkHttpProcessingTest : IDisposable
     {
-        #region Fields
-        private const string TestUrl = "http://www.microsoft.com/";
-        private const string TestUrlNonStandardPort = "http://www.microsoft.com:911/";
+#region Fields
+        private const string RandomAppIdEndpoint = "http://app.id.endpoint"; // appIdEndpoint - this really won't be used for tests because of the app id provider override.
```
Comment (cijothomas) [inline Src/DependencyCollector/Shared.Tests/Implementation/FrameworkHttpProcessingTest.cs]: good comments :)

File: CHANGELOG.md (hunk 18 above)
Comment (brahmnes) [inline CHANGELOG.md]: It's only for .NET46 and above, and not .NET core.
Sure, I can clarify

File: Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs [hunk 21]
```diff
@@ -64,7 +79,20 @@ public void TestHttpPostRequestsAreCollected()
...
+            {
+                OnSend = telemetry =>
+                {
+                    // The correlation id lookup service also makes http call, just make sure we skip that
+                    DependencyTelemetry depTelemetry = telemetry as DependencyTelemetry;
+                    if (depTelemetry != null && !depTelemetry.Data.StartsWith(FakeProfileApiEndpoint, StringComparison.OrdinalIgnoreCase))
```
Comment (Dmitry-Matveev) [inline Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs]: Minor: It looks like Data can be null if Dependency telemetry was initialized with default constructor. This particular test should not be affected.

File: Src/DependencyCollector/Shared.Tests/Implementation/FrameworkHttpProcessingTest.cs [hunk 22]
```diff
@@ -211,61 +229,275 @@ public void OnEndHttpCallbackSetsSuccessToFalseForMoreThan400()
...
-        /// Validates HttpProcessingFramework calculates startTime from the start of very first BeginGetRequestStream if any
+        /// Validates that even if multiple events have fired, as long as there is only
+        /// one HttpWebRequest, only one event should be written, during the first call
+        /// to OnEndHttpCallback.
+        /// </summary>
+        [TestMethod]
+        public void RddTestHttpProcessingFrameworkNoDuplication()
+        {
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(this.testUrl);
```
Comment (Dmitry-Matveev) [inline Src/DependencyCollector/Shared.Tests/Implementation/FrameworkHttpProcessingTest.cs]: Is it possible that customers reuse the same request on purpose, e.g. for some "ping" test or it will be returned from cache and they would need to create a new one each time to go via network? I'm asking to make sure we do not have an edge case here where the behavior should be the opposite (collect all, not the first one).

File: Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs (hunk 14 above)
Comment (cijothomas) [inline Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs]: currently rddf vs rddp identifies whether the dependency was collected with help of framework event source or via profiler. Since diagnostics source is a third mechanism (its different from framework eventsource) we need a way to know if the dependency was collected with it.

Comment (Dmitry-Matveev) [inline Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs]: I think they mean to add new value to RddSource enumeration like "ds" (after diagnostic source) and use it here instead of the RddSource.Framework.

File: Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs (hunk 19 above)
Comment (brahmnes) [inline Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs]: I tried to factor this out, but the lambda capture of sentTelemetry makes it hard to put it in a separate method.
Non-ideal, but leaving it this way may be the easiest.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 10 above)
Comment (Dmitry-Matveev) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: If it's not removed from 2.4 by stable milestone, I would recommend to implement `TryPrepare..()` in 2.4 context rather than wait for CoreFX, we would like to minimize # of the exceptions that are just thrown/swallowed especially on the "happy path".

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs [hunk 23]
```diff
@@ -0,0 +1,365 @@
...
+            // At any point, if the operation failed, it should just throw, including NullReferenceException.
+            // The caller should catch all exceptions and swallow.
+
+            // First step: Get all the reflection objects we will ever need.
+            Assembly systemNetHttpAssembly = typeof(ServicePoint).Assembly;
+            connectionGroupListField = typeof(ServicePoint).GetField("m_ConnectionGroupList", BindingFlags.Instance | BindingFlags.NonPublic);
+            connectionGroupType = systemNetHttpAssembly?.GetType("System.Net.ConnectionGroup");
+            connectionListField = connectionGroupType?.GetField("m_ConnectionList", BindingFlags.Instance | BindingFlags.NonPublic);
+            connectionType = systemNetHttpAssembly?.GetType("System.Net.Connection");
+            writeListField = connectionType?.GetField("m_WriteList", BindingFlags.Instance | BindingFlags.NonPublic);
```
Comment (Dmitry-Matveev) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: Do we need to assert for `ReflectionPermission` here before accessing `Nonpublic` fields?

File: Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs (hunk 19 above)
Comment (cijothomas) [inline Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs]: ok

File: Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs (hunk 14 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/FrameworkHttpProcessing.cs]: Got it, fixed.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 10 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: This file will certainly be removed from AI SDK.

File: Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs (hunk 21 above)
Comment (brahmnes) [inline Src/DependencyCollector/Net45.Tests/DependencyTrackingTelemetryModuleTest.cs]: Well, this is test code if the assumption wrong, it will fail and we will know. ;)

File: Src/DependencyCollector/Shared.Tests/Implementation/FrameworkHttpProcessingTest.cs (hunk 22 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared.Tests/Implementation/FrameworkHttpProcessingTest.cs]: .NET framework recommends that HttpWebRequest is not reused for multiple requests. In fact, this assumption was already in place in ProfilerHttpProcessing before this change, it used to look up the web request object from the cache and if it exists, it skips processing. This change doesn't change semantics.

File: Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs (hunk 23 above)
Comment (brahmnes) [inline Src/DependencyCollector/Shared/Implementation/HttpHandlerDiagnosticListener.cs]: It's better to just attempt the operation rather than checking it. If you have no access, it will just fail, and execution goes on. It falls back to just framework events without diagnostic source and no harm is done.

**Comment by SergeyKanzhelev** [comment]
@brahmnes please resolve the conflicts

**Comment by cijothomas** [comment]
Dont merge this until Functional tests are also updated.

**Comment by brahmnes** [comment]
This PR is replaced with https://github.com/Microsoft/ApplicationInsights-dotnet-server/pull/449
